File: time-attendance-selfservice-frontend/src/app/core/auth/auth.interceptor.ts
import { HttpInterceptorFn, HttpRequest, HttpHandlerFn, HttpErrorResponse } from '@angular/common/http';
import { inject } from '@angular/core';
import { catchError, switchMap, throwError, tap } from 'rxjs';
import { AuthService } from './auth.service';

export const authInterceptor: HttpInterceptorFn = (req: HttpRequest<unknown>, next: HttpHandlerFn) => {
  const authService = inject(AuthService);

  // Skip auth header for login and refresh endpoints
  if (req.url.includes('/auth/login') || req.url.includes('/auth/refresh')) {
    return next(req);
  }

  const token = authService.getAccessToken();

  // Add authorization header if token exists
  let authReq = req;
  if (token) {
    authReq = req.clone({
      headers: req.headers.set('Authorization', `Bearer ${token}`)
    });
  }

  console.log('[AuthInterceptor] Sending request:', req.method, req.url);

  return next(authReq).pipe(
    tap((event: any) => {
      if (event.type === 4) { // HttpResponse
        console.log('[AuthInterceptor] Response received for:', req.method, req.url, 'Status:', event.status);
      }
    }),
    catchError((error: HttpErrorResponse) => {
      console.error('[AuthInterceptor] Error for:', req.method, req.url, 'Status:', error.status);
      // Handle 401 unauthorized responses
      if (error.status === 401 && !req.url.includes('/auth/refresh')) {
        // Try to refresh token
        if (authService.getRefreshToken()) {
          return authService.refreshToken().pipe(
            switchMap(() => {
              // Retry original request with new token
              const newToken = authService.getAccessToken();
              const retryReq = req.clone({
                headers: req.headers.set('Authorization', `Bearer ${newToken}`)
              });
              return next(retryReq);
            }),
            catchError((refreshError) => {
              // If refresh fails, logout user
              authService.logout();
              return throwError(() => refreshError);
            })
          );
        } else {
          // No refresh token, logout user
          authService.logout();
        }
      }
      
      return throwError(() => error);
    })
  );
};

File: ess_mobile_app/lib/core/network/auth_interceptor.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../storage/secure_storage_service.dart';

/// Dio interceptor for handling JWT authentication.
/// Automatically adds Authorization header and handles token refresh.
class AuthInterceptor extends Interceptor {
  final Ref _ref;
  
  AuthInterceptor(this._ref);
  
  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) async {
    // Skip auth header for public endpoints
    final publicEndpoints = [
      '/api/v1/tenants/discover',
      '/api/v1/tenants/validate',
      '/api/v1/auth/login',
      '/api/v1/auth/refresh-token',
    ];
    
    final isPublicEndpoint = publicEndpoints.any(
      (endpoint) => options.path.contains(endpoint),
    );
    
    if (!isPublicEndpoint) {
      final accessToken = await SecureStorageService.instance.getAccessToken();
      if (accessToken != null) {
        options.headers['Authorization'] = 'Bearer $accessToken';
      }
    }
    
    handler.next(options);
  }
  
  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    // Handle 401 Unauthorized - try to refresh token
    if (err.response?.statusCode == 401) {
      final refreshed = await _tryRefreshToken();
      
      if (refreshed) {
        // Retry the original request with new token
        try {
          final accessToken = await SecureStorageService.instance.getAccessToken();
          err.requestOptions.headers['Authorization'] = 'Bearer $accessToken';
          
          final response = await Dio().fetch(err.requestOptions);
          return handler.resolve(response);
        } catch (e) {
          // Refresh failed, let the error pass through
        }
      }
      
      // Token refresh failed - clear tokens and redirect to login
      await SecureStorageService.instance.clearTokens();
      // TODO: Trigger logout/redirect to login
    }
    
    handler.next(err);
  }
  
  /// Attempt to refresh the access token using the refresh token.
  Future<bool> _tryRefreshToken() async {
    try {
      final refreshToken = await SecureStorageService.instance.getRefreshToken();
      if (refreshToken == null) return false;
      
      final tenantConfig = await SecureStorageService.instance.getTenantConfig();
      if (tenantConfig == null) return false;
      
      final dio = Dio();
      final response = await dio.post(
        '${tenantConfig.apiBaseUrl}/api/v1/auth/refresh-token',
        data: {'refreshToken': refreshToken},
      );
      
      if (response.statusCode == 200) {
        final newAccessToken = response.data['accessToken'] as String;
        final newRefreshToken = response.data['refreshToken'] as String?;
        
        await SecureStorageService.instance.saveAccessToken(newAccessToken);
        if (newRefreshToken != null) {
          await SecureStorageService.instance.saveRefreshToken(newRefreshToken);
        }
        
        return true;
      }
    } catch (e) {
      print('Token refresh failed: $e');
    }
    
    return false;
  }
}
